File: linked-list/reverse-linked-list/reverse-linked-list.ts
// Linked List: Reverse a singly linked list

class ListNode {
  val: number;
  next: ListNode | null;

  constructor(val: number, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function reverseList(head: ListNode | null): ListNode | null {
  let prev: ListNode | null = null;

  while (head !== null) {
    const next: ListNode | null = head.next;
    head.next = prev;
    prev = head;
    head = next;
  }

  return prev;
}

function printList(head: ListNode | null): void {
  const values: string[] = [];
  while (head !== null) {
    values.push(String(head.val));
    head = head.next;
  }
  console.log(values.join(" -> "));
}

function main(): void {
  const head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));

  const reversed = reverseList(head);
  printList(reversed);
}

main();

File: linked-list/reverse-linked-list/reverse-linked-list.cpp
// Linked List: Reverse a singly linked list
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int v) : val(v), next(nullptr) {}
};

ListNode* reverseList(ListNode* head) {
    ListNode* prev = nullptr;
    while (head != nullptr) {
        ListNode* next = head->next;
        head->next = prev;
        prev = head;
        head = next;
    }
    return prev;
}

void printList(ListNode* head) {
    while (head != nullptr) {
        cout << head->val << (head->next ? " -> " : "\n");
        head = head->next;
    }
}

int main() {
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);

    ListNode* reversed = reverseList(head);
    printList(reversed);

    return 0;
}
